core/include/assert.h

```diff
@@ -25,11 +28,55 @@
 extern "C" {
 #endif
 
+#ifdef DOXYGEN
+/**
+ * @brief   Activate verbose output for @ref assert() when defined.
+ *
+ * Without this macro defined the @ref assert() macro will just print the
+ * address of the code line the assertion failed in. With the macro defined
+ * the macro will also print the file, the code line and the function this macro
+ * failed in.
+ *
+ * To define just add it to your `CFLAGS` in your application's Makefile:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.mk}
+ * CFLAGS += -DDEBUG_ASSERT_VERBOSE
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ */
+#define DEBUG_ASSERT_VERBOSE
+#endif
+
 /**
  * @brief   the string that is passed to panic in case of a failing assertion
  */
 extern const char assert_crash_message[];
 
+#ifdef NDEBUG
+#define assert(ignore)((void) 0)
+#elif defined(DEBUG_ASSERT_VERBOSE)
+#include <stdio.h>
+
+/**
+ * @brief   Information about the location of a failed assertion
+ * @note    This was introduced for @ref _assert_failure() for memory size
+ *          optimization
+ */
+typedef struct {
+    const char *file;   /**< file name of the file the assertion failed in */
+    unsigned line;      /**< code line of the file the assertion failed in */
+} _assert_failure_info_t;
```

Comment:
The rational is to carry the failure information (filename and line of assertion failure) to the `_assert_failure()` function. Additionally this allows the compiler to have the filename `file` be only provided in memory once per compile unit (which makes sense, because it is literally the compile units source code name), since we declare this variable as `static const`. The idea came from @Kijewski (see https://github.com/RIOT-OS/RIOT/pull/5881#discussion_r81353720)